### ISSUE TITLE
A required field is allowed to save empty when page is already published.

### ISSUE THREAD

**Issue by somatonic** [ryancramerdesign/ProcessWire#1743]
First time I've seen that this is possible. I assumed you can't empty a field that is required on a published page and save it. Means the field is empty after saving and shows an error. This caused the site to not work anymore.

It should never be possible to empty and save a field that is required.

**Comment by isellsoap** [comment]
Maybe related: https://github.com/ryancramerdesign/ProcessWire/issues/1121 and https://github.com/ryancramerdesign/ProcessWire/issues/1378.

**Comment by adrianbj** [comment]
On the other side of the coin, I think it should be possible to save an unpublished page with an empty required field. It shouldn't be checked until published. Maybe this should be a configurable option. I mentioned this somewhere else, but can't find where at the moment.

**Comment by somatonic** [comment]
It's already possible to save an unpublished page with an empty required value. It just can't be published until it's populated.

In my opinion it should save other fields but not empty the required field and stop functionality stop working depending on its value.

**Comment by adrianbj** [comment]
> It's already possible to save an unpublished page with an empty required value. It just can't be published until it's populated.

Right sorry - I was forgetting the situation - I think maybe I asked if the required warning would show up until you try to publish.

**Comment by teppokoivula** [comment]
@somatonic: in some old posts Ryan pointed out that unpublishing pages if a required field is cleared later on could be considered a problem, so perhaps this was intentional. Hard to say, but either way: in my opinion clearing a required field should definitely unpublish the page.

**Comment by somatonic** [comment]
@teppo, no not unpublish the page, but simply not save the empty value.

**Comment by teppokoivula** [comment]
@somatonic From what I've read so far, that seems to be even more problematic. In fact Ryan has suggested that it can't be done, and mentioned a JavaScript-based solution (well, hack) as an alternative.

Though I do agree that "don't save the page unless all required fields are populated" would be the best option overall, unpublishing published pages if they are saved and a required field is left empty would be the most straightforward approach, and it would be consistent with current behaviour for unpublished pages.

**Comment by ryancramerdesign** [comment]
One of the considerations with a system like PW are that with flexibility comes responsibility. PW operates on the premise that a page structure is never "final", and could have new fields added to it at any time. You might have a template used by a thousand pages, and you add a new field to it and make it required. What's supposed to happen to those 1000 pages? Any low-level "required" state creates a scenario that can't be met if a system is to remain flexible and scalable. That's why "required" is something for the editor, and not something one can ever expect at the API level. 

What it comes down to is that you should never write code that makes assumptions about the content of a value. What you *can* expect on the API level is what are indicated as possible return values and types. Meaning, if you've got a Page field and chosen that you want it to always return a PageArray, then it will always return a PageArray (whether it has items in it or not). So it's okay to make assumptions about type, but not content. Type(s) are fixed, content is not. Something like empty vs non-empty can occur with or without editor error. So it's something you'd have to code for regardless of "required" state or what happens on the admin UI side.

**Comment by isellsoap** [comment]
Making sure your code doesn’t make assumptions about the content of a value is one thing.

But with the current behavior the required state of a field actually has no real functionality behind it (I’m looking at the scenario @somatonic describes in the first post of this issue). One can empty a filled out and required text field, save the page and functionality-wise it’s like the required state wasn’t there in the first place because the user could actually save the page and alter the value of the field (making it empty). The only this that’s happened is a optical hint that something went wrong.

**Comment by somatonic** [comment]
PW makes the assumption here that "it's ok to empty that value" even though it's required. No reason and logic in that really.

You also can't empty the page "name" field, cause it's a system field that is required for the system to work. It will show an error and leave the previous value. I want to be able to have a custom field that works the same way when required. Maybe we need an option to "enforce" and not allow empty value getting saved on a published page.

When I build a system or app with PW, I want to make sure there's a value. When I personally would add a field to a page and it has to be required, I as the developer have to make sure it has a value filled before I make it dependent on that value and making it "live". So I could also imagine I would unpublish all pages and let editor fill in and publish them.

**Comment by ryancramerdesign** [comment]
@isellsoap That's correct, that is the scope and purpose of the required state. PW2 didn't originally come with a required option at all for the reasons mentioned. But folks wanted something to at least help the editors along, so that's what it is. Though I always worried it would create a false expectation.

Perhaps when we added the block publish of an unpublished page that is missing required fields, that may have created the illusion. I think doing what Teppo suggested (unpublishing a page automatically for missing a requirement) sounds interesting, but bigger picture has more potential to break a site, than writing code with content dependencies does... at least on significant pages in the site (homepage, top-level, etc.). 

Worth nothing though is that ProDrafts opens the door to a lot more possibilities here, as we can potentially make it force a save to a draft rather than published if a requirement isn't fulfilled. Though that still doesn't change the underlying consideration of coding for type not content. I'm reluctant to keep building on top of the illusion because it makes people code for the illusion rather than the reality (flexibility) that any page can exist with an empty value in any field. But I'll definitely explore it further in ProDrafts, since it does open the door to a lot more in this regard.

**Comment by ryancramerdesign** [comment]
@somatonic a page can't actually exist in the DB with a name field, since it is one of the unique keys in the MySQL index. Whereas the same is not true of custom fields. But if you have a need to block save of a page for a specific empty field, you could always do it with a hook throwing a WireException in your 

/site/templates/admin.php file:
`````````
$wire->addHookBefore('Fieldtype::savePageField', function($event) {
  $page = $event->arguments(0);
  $field = $event->arguments(1); 
  $value = $page->get($field->name); 
  if($field->required && empty($value)) {
    throw new WireException("$field->name is a required field!"); 
  }
}); 
`````````
There may be some bad side effects here, so test on a dev server not a live one. However, it shoudln't prevent other fields from being saved, as the exception is caught and displayed for each field that is saved. Also you may want to update the if() condition to focus on specific fields to limit the potential scope of issues. Again something like this just furthers the illusion, so don't let it substitute for code.

**Comment by isellsoap** [comment]
> Though I always worried it would create a false expectation.

Well, than you should change the description of the required state, because it says

> If checked, a value will be required for this field.

That’s pretty unambiguous to me. :-)

> I think doing what Teppo suggested (unpublishing a page automatically for missing a requirement) sounds interesting […]

I think that would create a horrible user experience, to be honest.

> You might have a template used by a thousand pages, and you add a new field to it and make it required. What's supposed to happen to those 1000 pages?

Nothing. Everything should work like it works now.

Obviously I don’t know the technical implications of such a change but it seems to me that they are rather severe seeing that you do not want to change the required state (or heavily argue against it). I actually am really interested into why it is such a difficult thing to tackle. For example I don’t see why such a change would have negative impacts on the great flexibility of ProcessWire.

**Comment by ryancramerdesign** [comment]
The issue report here is about a page that apparently broke because a field
was empty when it was expected to be populated. I've communicated it is
impossible to assume that something will be populated, regardless of
required state. That's because the requirements for fields, presence of
fields, and content of fields can change outside of the page editor. Pages
are not one-time submission forms. The schema can change without an editor.
Fields can be added, removed, requirements can change, and so on, while the
content of that page does not change. So the fact is one can never make an
assumption of content. Given that fact, a "required" field cannot guarantee
that a value will always be present. There is no "difficult thing to
tackle" here because there is no thing to operate upon. It's what a user
accepts in exchange for the flexibility of being able to alter schema to
existing pages, templates and fields. Does that make sense? The more one
starts to assume something about a value beyond its type, the greater
potential for code that doesn't reflect reality. So the question should
instead be, how far to carry that illusion? In my opinion, not very far. My
preference was to leave the concept out of the system entirely, since it
inevitably leads to the content of this issue report. But I tried to meet
the request in the middle some years ago. Granted this is all "big picture"
and maybe it's hard to grasp through the lens of an individual site.
Perhaps an individual installation would be perfectly fine using a hook
like the one I wrote to Soma above. And maybe that would solve the specific
problem he ran into. But in the big picture I think it's better that people
code for the nature of the system, where schema and requirements can
change, which is a good thing.

**Comment by isellsoap** [comment]
> The issue report here is about a page that apparently broke because a field was empty when it was expected to be populated. I've communicated it is impossible to assume that something will be populated, regardless of required state. That's because the requirements for fields, presence of fields, and content of fields can change outside of the page editor.

I totally agree with you and that’s also my point of view. 

> Pages are not one-time submission forms.

When a editor edits a page in the GUI and saves the page, than that's a one-time submission form. 

> The schema can change without an editor.

That's correct but I don't see what that has to do with my argument. What I find troubling is the fact that currently a schema _that is the status quo at the time of a page save in the GUI_ doesn’t get correctly validated (required and filled-out field can be empty and the page gets saved anyway **with the empty value of the field**). That's the essence of my argument.

What I would expect is simply that in such a scenario the value of the previous state of the field should be re-populated with the appropriate error message. That’s all. That would eliminate most weird use cases I (and others) described but still wouldn’t restrict or limit PW’s flexibility at all.

**Comment by teppokoivula** [comment]
Unless I'm missing something here, I think this would be the most straightforward approach, and wouldn't change much in comparison to how things currently work:

- Since ProcessWire supports unpublished state, another option is to save all fields, but don't make the page published until all required fields are populated. I believe this is already how it works.
- If a page has already been saved and required fields are added, or fields are made required, don't do anything. That's all there is to it, in my opinion; no need to make this any more complicated than it is.
- Next time the page is saved, if it still has empty required fields, unpublish it.

This means that you *still* can't expect a page to have values in all required fields, but it also means that the required state works for the most important purpose it has: user can't publish any *new* pages without populating all required fields. If a field is made required or required field is added later on, in my opinion it's the developer's responsibility to make sure it doesn't cause issues (and fix existing data if needed).

The more "hard-core" approach would be to change two parts of this:

- If a page is saved and it has empty required fields, just don't. Return to page editor with an error message. This is bad for usability and probably not what anyone wants.. except if combined with client-side caching solution, in which case it could work pretty well.
- Additionally (and this I believe is what @somatonic suggested) you could make it impossible to clear required fields by returning them to their previous populated state if they're empty during page save.

In some ways I think we're making this all too complicated. At least in my use cases schema changes to existing pages are always kind of a big deal, and I don't mind making sure that content is up to date with them. ProcessWire doesn't have to account for everything, but in my opinion should make it easy (or at least possible) to account for most (potentially breaking) actions of regular users.

**Comment by LostKobrakai** [comment]
I want to add my support for simply making it impossible to empty out required fields, which were previously filled. Any schema changes to pages are in the responsibility of the dev. If pages do need some fields to be filled after a change it's their job to make sure these requirements are fulfilled. But as soon as a required field is filled and the field is required it should not be possible to empty that thing again via the page editor. Saving via the api is a whole other context, which I think is fine in their current form.

**Comment by isellsoap** [comment]
@teppokoivula 

> make it impossible to clear required fields by returning them to their previous populated state if they're empty during page save and if a previous value exists.

Why is that the more "hard-core" approach? I think that’s the only reasonable approach and the most user-friendly one. I think that is the go-to solution.

Regarding your comments about the unpublishing of pages

> Next time the page is saved, if it still has empty required fields, unpublish it.

and

> Otherwise unpublish the page.

I strongly disagree. Who would expect such a behavior? I think that would be a very bad user experience. I think that would be the "hard-core" solution.

**Comment by teppokoivula** [comment]
@isellsoap Keeping pages unpublished until all required fields are filled in is already built-in feature. "Real required fields", i.e. "you can't clear this required field once the page is published, period" and "you can't save this page without filling in all required fields, not even as an unpublished one" would be a relatively big change. Perhaps "strict" would be more fitting than "hard-core".

Since we're talking about strong disagreements here, I strongly disagree with your point about unpublishing page being unexpected. This would be iin line with current "can't publish a page with empty required fields" behaviour. In my opinion the only *completely unexpected* thing here is that you can currently clear those fields after a page has been published *and it stays published.*

The "very bad user experience" point is also questionable, since It's actually very easy to notice if a page was unpublished. Combined with a proper notice about what just happened I think that's actually much better for user experience than the site going wonky because a field that was expected to have a value suddenly doesn't.

Whatever route we go, it seems to me that this should probably be configurable behaviour :)

**Comment by isellsoap** [comment]
> I disagree with your point about unpublishing page being unexpected. This would be in line with current "can't publish a page with empty required fields" behaviour.

I thought more about the impacts that unpublishing an already published page has for visitors of a website. Not publishing a not-yet-published page doesn’t concern visitors of a website at all, because they couldn’t interact with the page yet. They don’t know it exists because it was never published. This is definitely a non-critical behavior.

Unpublishing a already published page after clicking a "save" button on the other hand definitely is a critical behavior:

a) It is just plain and simple a bad user experience when you hit a button that says "save" and something completely unexpected happens. From a semantic point of view I expect that the _published_ page gets saved, and not that the page gets the opposite state it currently is in.

b) Visitors immediately get a 404 error when they visit the URL of a page. I think I don’t have to go into the details into how this is not good.

So in that regard I think saying the two approaches "are in line" with each other is wrong, isn’t it?

**Comment by teppokoivula** [comment]
> So in that regard I think saying the two approaches "are in line" with each other is wrong, isn’t it?

I get your point. Looks like we were talking about user experience from slightly different points of view: I was thinking that the user experience from the point of view of content editor would be just fine, but it's true that a page disappearing from the front-end can be confusing for the end-users browsing the site. Whether 404 is better or worse than empty required fields depends entirely on the case at hand.

Just so that we don't get too distracted here:

- In my opinion the best option by far would be proper required fields, i.e. a page can't be saved unless all required fields are filled in and required fields can't be emptied once the page exists.
- If we can't have required fields that are actually required, in my opinion the page should definitely be unpublished once those fields are cleared. It doesn't make any sense that I can't publish a page with empty required fields *but I can fill them in, publish the page, and then clear those fields*.

In other words I agree with you on all points except that I see unpublishing the page as a better solution than the current behaviour. I don't see the published/unpublished state change as such a huge thing, and in my opinion "save" unpublishing the page *if you've just cleared fields marked as required* can't be wildly unexpected behaviour.

Adding a front-end check and a warning message (something along the lines of "required fields are empty, the page will be unpublished if you save – do you want to continue anyway?") might be a good idea, but while we're at it, we might as well add "required" attributes to further enhance this. If the user deliberately ignores these hints and/or somehow overrides them, unpublishing the page should be fine.

**Comment by isellsoap** [comment]
Full acknowledgement! Especially your last paragraph has cool ideas in it. Would be awesome to see these additions in the core!

**Comment by matjazpotocnik** [comment]
So what we have now:

1. "It should never be possible to empty and save a field that is required. Emptied fields should revert to previous values." 

2. "a page can't be saved unless all required fields are filled in and required fields can't be emptied once the page exists" (@teppokoivula)

Let suppose that we have "proper required fields". If we ignore javascript and browser checking of required fields, what should happen after hitting save button? Should emptied fields revert to previous values?

I'm for proper required fields. I'm against "automatic" unpublishing, unless user confirms.

**Comment by apeisa** [comment]
I think this issue is getting so big, that it would be shame to be left out of fun :)

I understand the points Ryan is pointing out at the API level, so I think focus here should be what default **admin page edit view** does in situation where already published page is saved again, now with a required field missing a value.

There is also two different situations and I think we should have same way to handle both situations (I like simple solutions). One is where editor tries to empty value in required field and second one is where there is new required field added into template (or existing field made required). 

Unpublishing an already published page is no go in my opinion. How and why did that happen? I just don't understand the logic here. 

What @isellsoap suggest is best option in my opinion. Saving all other fields and repopulating the empty required field with the old value (with error message noticing about that). Same thing if the required field was already empty (it stays empty with error message, but no page gets unpublished and all other values do get saved normally).

**Comment by teppokoivula** [comment]
> Unpublishing an already published page is no go in my opinion. How and why did that happen? I just don't understand the logic here.

The logic here is quite simple, really: if a page can't be published until all required fields are filled in, then why is it that it can it exist in a published state if those fields are later cleared? In other words: if filling in required fields is a prerequisite for publishing a page, in my opinion it's only logical that clearing those fields also reverts the page back to unpublished state :)

That being said, I don't disagree with you: reverting to previous values would be a sensible option. After thinking about this a bit more, I can't help feeling that if I've just cleared a value from a field, I'd be a bit surprised to still see it after saving (how and why did that happen?) but at this point I'm pretty sure that I'm just overthinking this.

Either way a sensible explanation is required: what happened and why. The user should see a clear and helpful message explaining that "the page was unpublished because you cleared one or more required fields", or alternatively something along the lines of "you tried to clear required fields, old values were restored to fields field-a, field-b, and field-c".

In my opinion the only 100% obvious behaviour would be forcing required state so that you can't submit the page edit form without filling all required fields in the first place; anything other than that should be combined with some kind of notification about what just happened.

**Comment by Renobird** [comment]
> The logic here is quite simple, really: if a page can't be published until all required fields are filled in, then why is it that it can suddenly exist in published state if those fields are cleared later on?

I agree with this. Especially if there have been changes made that require a field to have a value, then you should not be able to resave a page in a published state (from the admin at least) without it being set. If I add a new required field to a template, then I want that field to be filled in the next time the page is saved, or unpublished the page and notify the user.

Notifications are the key. As long as it's clear what happened and why, this seems like the most expected behavior. 

I would vote that this be the default behavior, and any alternative be a configurable option.

**Comment by matjazpotocnik** [comment]
> The logic here is quite simple, really: if a page can't be published until all required fields are filled in, then why is it that it can suddenly exist in published state if those fields are cleared later on?

I see your point. Consider this: if you add required field to the template later on, all published pages using that template are from now on in the "wrong state", aren't they? Should they get unpublished? 

If you add required field to the database table (general speaking) and then you edit record, all applications I have seen does not allow you to save the record, until all required fields are filled in. If you click cancel during the edit, the record is in the same state as it was before edit, the record is not changed into "unpublished" state. It also does not save other fields you may have changed. Its all or nothing approach.

> I can't help feeling that if I've just cleared a value from a field, I'd be a bit surprised to still see it after saving (how and why did that happen?)

I totally agree. If user clears required field and hit Save, notice is echoed about required field, another notice just above the required field is saying "Missing required value" but the value is there!

**Comment by ryancramerdesign** [comment]
Sorry I'm late getting back here, I lost track of this thread. I can see there's a lot of opinions on this. But there's just no getting around the reality, like I tried to explain earlier. The nature of the system means it's always going to be possible for there to be empty required fields, regardless of anything else. So if someone is coding a site making assumptions like something is always going to be populated, then sooner or later they are going to get bitten, regardless of what strategies we might implement (like those mentioned above). Given that, I think the only truly correct behavior is the current one. And the only real solution is to write code in a manner that reflects the reality of the system–that empty values can exist. It's not a matter of opinion– So long as templates, fields and fieldgroups are to remain flexible and open to changes, that is the reality. But once that is understood, I think it's okay to try and build some useful facade on top of it, for those that helpful. And it sounds like there are some cases where it would be. I've tried to accommodate the requests above in the latest commit linked here, by making the option configurable at the template level. 

Edit a template, click the Advanced tab, and you'll see these options for the field "Required field action": 

- Alert the user (default)
- Restore previous value (when available)
- Unpublish the page (when allowed)

The only option you can always 100% count on is the existing behavior (the first line listed) because there may not be a previous value to restore, and the user might not have permission to unpublish the page (or the page may be of a type that can't exist unpublished), or there may be other factors. That's why the options are qualified with "when available" and "when allowed". So OP of this thread will still have to code for the possibility of empty values, regardless of which option is chosen, or else they'll encounter the same error. But I think you guys understand that code side of it, so this discussion is really more about the page editor than the original issue report. And in that context, I think these new options should be helpful. 

While the "unpublish the page" option might sound like a no-go for many websites (especially primary pages), I think it's got good use cases on other types of web apps and perhaps on large inventories [in a website] where a page getting unpublished isn't going to take down a major section of the site. That's why I opted to make these configurable at the template level, because it really seemed like there are places where you might one one behavior or another. Maybe that extends to configuration at the field level too, but we'll start here for now.

**Comment by teppokoivula** [comment]
Thanks, @ryancramerdesign. It's definitely good to keep in mind that this has been discussed from two points of view here: one is about always enforcing values for all fields (which we all know is impossible) and the second is about being helpful towards and/or enforcing certain guidelines for content editors (which is not only possible but, at least in my opinion, also very useful).

**Comment by somatonic** [comment]
I have nothing to add here, as my point of view stands. I completely understand what's the thinking behind this is. I still can't agree that the current behavior is "correct".

The issue I had was that the date of pages was used to find them and it would also return pages that had empty date. So the solution what to add a date!='' to the selector. 

Thing is, that even if there's correct coding, the page maybe depend on a value to have a certain function going on that would be broken if there's no value. By "broken" I mean not functioning as it should (the page still shows). All we could do is code it that there's a message: "Hey a required value isn't there anymore...". So the visitor get's a "broken" page, the editor only sees his website isn't functioning correctly anymore. To prevent that we have a "required" feature, that turns out isn't preventing from getting cleaned by an editor. 

Thanks very much Ryan for the options. Very welcomed.

As I see it the second option is the only correct one. Currently, when PW doesn't allow me to publish a page without filling in the required field, there will ALWAYS be a previous value, except if the data changed from other sources, (which is not the editor fault), but then the field would be empty already when editing the page.
